Add support for dependencies on app targets from different projects

TamarMilchtaich opened the pull request:
In certain cases, such as when the test host app resides in a different project, it's necessary to depend on an .app target from another project. To enable this, a file reference to the app is required, followed by setting up a dependency on it. This commit introduces file references for all .project executable targets, and adds a build phase to copy the .app file, making it available during the build process.

### Short description 📝

Allow dependency of `.app` targets from another project.

### How to test the changes locally 🧐

Create a project with an app target, and depend on this target in another project. Before the change, there should be no affect, and after the change, there should be a dependency.

For a more realistic flow, create a test host app, and use it in a project with on app targets (fixtures that demonstrate this - TBD)

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

More tasks:

- [x] Check if this solves this issue: https://github.com/tuist/tuist/issues/6130
- [x] Add fixture with such example

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

fortmarek commented on Sources/TuistCore/Graph/GraphTraverser.swift:
```diff
@@ -616,6 +621,18 @@ public class GraphTraverser: GraphTraversing {
         return Set(dependencies)
     }
 
+    public func executableDependencies(path: Path.AbsolutePath, targetName: String) -> Set<GraphDependencyReference> {
```
Is this extra method necessary or can the `.app` product be included as part of the return value of [copyProductDependencies](https://github.com/tuist/tuist/blob/4d879469bdc81d1a5b962d22d28c4b3126a2a180/Sources/TuistCore/Graph/GraphTraverser.swift#L602)?

Also, should we include the executables in the build phase only when the target is of type of UI or unit test? I don't think it'd make sense to copy the `.app` target to `.framework`, would it?

TamarMilchtaich replied:
1 -> I'll remove it and add to `copyProductDependencies` :)
2 -> I think it can also be useful in other cases (still want to make sure, but maybe also solves this: https://github.com/tuist/tuist/issues/6130)
In general, there are two approaches that can be taken here:

A) tailor the solution for for `.app` dependencies in a `unit/uiTest` targets
B) use a more general solution, where general dependencies on apps from another project will be supported

The thinking with the latter is that it will probably cover more cases where an app from another target is used as a dependency, but maybe we rather add each of these cases individually to make it safer. What do you think?

fortmarek replied:
> The thinking with the latter is that it will probably cover more cases where an app from another target is used as a dependency, but maybe we rather add each of these cases individually to make it safer. What do you think?

I'd lean to A) unless the changes here also solve https://github.com/tuist/tuist/issues/6130. It's usually better to minimize the behavior change to what one is trying to fix. It's often difficult to foresee all the implications of updating foundational logic in the generation process.

fortmarek replied:
That being said, would you mind checking if we can also fix https://github.com/tuist/tuist/issues/6734 with some minimal changes in this PR? That issue seems quite related.

TamarMilchtaich replied:
So:
1. Re https://github.com/tuist/tuist/issues/6130 -> the app in the example project doesn't reference the watch app, but since it's an `.app` it should be the same, and indeed - after the change in this PR, the watchapp is embedded as expected
2. Re https://github.com/tuist/tuist/issues/6734 -> it's pretty similar indeed, here there is an appExtension embedding phase (`generateAppExtensionsBuildPhase`), but it calls `graphTraverser.appExtensionDependencies`, which only considers local dependencies. So I also added the appExtension references in the graph traverser, and changed `extensionKitExtensionDependencies` to look at all dependencies. It's probably the same for the graph traverser's `extensionKitExtensionDependencies` and `appClipDependencies`, but for the sake of targeted changes - only changed `appExtensionDependencies` for now.
For ease of review, added all the changes in a new commit (to be squashed later)

fortmarek reviewed (commented):
Hey @TamarMilchtaich 👋 

Just a minor comment, but I'm generally aligned with the direction you're taking here 🙌

fortmarek commented on Sources/TuistCore/Graph/GraphTraverser.swift:
```diff
@@ -611,11 +626,26 @@ public class GraphTraverser: GraphTraversing {
         return Set(dependencies)
     }
 
+    public func copyExecutableDependencies(
```
```suggestion
    public func executableDependencies(
```

fortmarek reviewed (commented):
I'm generally aligned with the approach here!

It will be very useful if you created a fixture in this [directory](https://github.com/tuist/tuist/tree/main/fixtures) that does not work in the latest Tuist release but it does work with the changes from your branch.

You can then test this feature E2E using an acceptance test: https://github.com/tuist/tuist/blob/main/Tests/TuistGeneratorAcceptanceTests/GenerateAcceptanceTests.swift

Let me know if you need any extra pointers!

TamarMilchtaich commented:
@fortmarek Added fixture with acceptance test :) The test passes after the change in this PR, and fails before it

The fixture includes dependency on three app targets from another project, which should correspond to the three issues fixed (the watch app target now showing app and test host target not showing up are the same issue, but I thought we would benefit from having both examples)

fortmarek commented on Sources/TuistCore/Graph/GraphTraverser.swift:
```diff
@@ -207,6 +212,16 @@ public class GraphTraverser: GraphTraversing {
         }
     }
 
+    func executableNonLocalDependencies(path: Path.AbsolutePath, targetName: String) -> Set<GraphDependencyReference> {
```
```suggestion
    private func executableNonLocalDependencies(path: Path.AbsolutePath, targetName: String) -> Set<GraphDependencyReference> {
```

TamarMilchtaich commented:
Added tests and the fixes from above, and squashed. PLMK if you think there are more cases we need to test

I also wanted to add what we discussed here:
https://github.com/tuist/tuist/pull/6895#issuecomment-2441006062
But actually it's problematic when the test host is non-local -
- When it's local, taking the first app target will make sense, and probably give the main app to run on
- But with non-local, there might be many app targets, only one of those is the test host. The way the user determines which one of those is the test host is actually by manually setting those settings

So it makes sense to me to keep it as-is, but if there's some other solution that you prefer - we can change it

fortmarek reviewed (approved):
Thanks a lot again for this PR! We can merge once the CI is green. Let me know if you need help with resolving the failing test case.

TamarMilchtaich commented:
Yay! Thank you for reviewing :)

The failing test seems to pass locally (even after rebase over `main`), and as much as I can tell doesn't seem much related to the changes here. Any suggestion about how to recreate the fail? And could it be caused by some flakiness in the test that's unrelated to the changes here?

fortmarek commented:
> Any suggestion about how to recreate the fail? And could it be caused by some flakiness in the test that's unrelated to the changes here?

Updating with the latest changes from `main` has resolved the issue 🙂

TamarMilchtaich commented:
Thanks! Merging :)

fortmarek commented:
Yes! We're in a middle of a release, so I'm just waiting for that to finish as it makes the commits easier to follow. This PR will be included in the next one that will be released some time next week 🙂

fortmarek commented:
Thanks again for this contribution!
